pdr_backend/sim/sim_plotter.py

```diff
@@ -366,6 +367,100 @@ def _add_subplot_log_loss_vs_time(self, fig, row):
         )
         fig.update_yaxes(title_text="log loss", row=3, col=1)
 
+    @enforce_types
+    def plot_prediction_residuals(self):
+        # set titles
+        aim = self.st.aim
+        s1 = "Residual vs time"
+        s2 = "Residual distribution"
+        s3 = "Residuals normal Q-Q"
+        s4 = "Residuals correlogram"
+
+        # make subplots
+        fig = make_subplots(
+            rows=2,
+            cols=2,
+            subplot_titles=(s1, s2, s3, s4),
+            vertical_spacing=0.18,
+        )
+
+        # fill in subplots
+        self._add_subplot_residual_vs_time(fig, row=1, col=1)
+        self._add_subplot_residual_pdf_nq(fig, row=2, col=1)
+        #self._add_subplot_residual_correlogram(fig, row=2, col=2)
+        
+        return fig
+
+    @enforce_types
+    def _add_subplot_residual_vs_time(self, fig, row, col):
+        y = self.st.aim.yerrs
+        x = list(range(len(y)))
+        fig.add_traces(
+            [
+                # points: residuals vs time
+                go.Scatter(
+                    x=x,
+                    y=y,
+                    mode="markers",
+                    marker={"color": "black", "size": 2},
+                    showlegend=False,
+                ),
+                # line: horizontal error = 0
+                go.Scatter(
+                    x=[min(x), max(x)],
+                    y=[0.0, 0.0],
+                    mode="lines",
+                    line={"color": "grey", "dash": "dot"},
+                    showlegend=False,
+                ),
+            ],
+            rows=[row]*2,
+            cols=[col]*2,
+        )
+        fig.update_xaxes(title="time", row=row, col=col)
+        fig.update_yaxes(title="residual", row=row, col=col)
+
+        return fig
+
+    @enforce_types
+    def _add_subplot_residual_pdf_nq(self, fig, row, col):
+        x = self.st.aim.yerrs
+        add_pdf(fig, x, row, col, "residual")
+        #add_pdf_nq(fig, x, row, col, "residual")
+        
+        return fig
+
+    @enforce_types
+    def _add_subplot_residual_normal_qq(self, fig, row, col):
+        y = self.st.aim.yerrs
+        x = list(range(len(y)))
+        fig.add_traces(
+            [
+                # points: residuals vs time
+                go.Scatter(
+                    x=x,
```

Comment:
This code seems to be duplicated, maybe it can be abstracted into a special type of figure in the future. I know this PR is not done yet, but I'm leaving this as a possibility.

Reply:
Fyi I will be fixing the duplicate code. As you know, I hate dry violations :)